pkg/controller/datasource.go

```diff
@@ -0,0 +1,198 @@
+	dataSrc := sdk.Datasource{
+		OrgID:     uint(ds.Spec.OrgID),
+		Name:      ds.Spec.Name,
+		Type:      ds.Spec.Type,
+		Access:    ds.Spec.Access,
+		URL:       ds.Spec.URL,
+		IsDefault: ds.Spec.IsDefault,
+	}
+
+	err := c.setGrafanaClient(ds.Namespace, ds.Spec.Grafana)
+	if err != nil {
+		return err
+	}
+
+	if ds.Status.DatasourceID != nil {
+		err := c.updateDatasource(ds, dataSrc)
+		if err != nil {
+			return errors.Wrapf(err, "can't update Datasource, reason: %v", err)
+		}
+		return nil
+	}
+	statusMsg, err := c.grafanaClient.CreateDatasource(context.TODO(), dataSrc)
+	if err != nil {
+		return err
+	}
+	glog.Infof("Datasource is created with message: %s\n", pointer.String(statusMsg.Message))
+	_, err = util.UpdateDatasourceStatus(context.TODO(), c.extClient.GrafanaV1alpha1(), ds.ObjectMeta, func(st *api.DatasourceStatus) *api.DatasourceStatus {
+		st.Phase = api.DatasourcePhaseSuccess
+		st.Reason = "Successfully created Grafana Datasource"
+		st.ObservedGeneration = ds.Generation
+		st.DatasourceID = pointer.Int64P(int64(pointer.Uint(statusMsg.ID)))
+
+		return st
+	}, metav1.UpdateOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to update Datasource phase to %q\n", api.DatasourcePhaseSuccess)
+	}
+	return nil
+}
+
+func (c *GrafanaController) updateDatasource(ds *api.Datasource, dataSrc sdk.Datasource) error {
+	dataSrc.ID = uint(pointer.Int64(ds.Status.DatasourceID))
+
+	statusMsg, err := c.grafanaClient.UpdateDatasource(context.TODO(), dataSrc)
+	if err != nil {
+		return err
+	}
+	glog.Infof("Datasource is updated with message: %s\n", pointer.String(statusMsg.Message))
+	// Update status to Success
+	_, err = util.UpdateDatasourceStatus(context.TODO(), c.extClient.GrafanaV1alpha1(), ds.ObjectMeta, func(st *api.DatasourceStatus) *api.DatasourceStatus {
+		st.Phase = api.DatasourcePhaseSuccess
+		st.Reason = "Successfully updated Grafana Datasource"
+		st.ObservedGeneration = ds.Generation
+
+		return st
+	}, metav1.UpdateOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to update Datasource phase to %q\n", api.DatasourcePhaseSuccess)
+	}
+	return nil
+}
+
+func (c *GrafanaController) runDatasourceFinalizer(ds *api.Datasource) error {
+	newDS, err := util.UpdateDatasourceStatus(context.TODO(), c.extClient.GrafanaV1alpha1(), ds.ObjectMeta, func(st *api.DatasourceStatus) *api.DatasourceStatus {
+		st.Phase = api.DatasourcePhaseTerminating
+		st.Reason = "Terminating Datasource"
+		st.ObservedGeneration = ds.Generation
+
+		return st
+	}, metav1.UpdateOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to update Datasource phase to %q\n", api.DatasourcePhaseTerminating)
+	}
+	ds.Status = newDS.Status
+
+	if ds.Status.DatasourceID == nil {
+		return errors.New("datasource can't be deleted: reason: Datasource ID is missing")
+	}
+	dsID := uint(pointer.Int64(ds.Status.DatasourceID))
+	statusMsg, err := c.grafanaClient.DeleteDatasource(context.TODO(), dsID)
+	if err != nil {
+		return err
```

Comment:
Record error event

Reply:
Record error events for Grafana datasource create/update/delete events